### PR TITLE
feat(query-engine-wasm): vertical slice

### PR DESCRIPTION
**Pull request by jkomyno**
This PR is based on https://github.com/prisma/prisma-engines/pull/4456, and contributes to https://github.com/prisma/team-orm/issues/549 (but it's not feature complete yet, so it doesn't close it).

This is the source code of [`@prisma/query-engine-wasm`](https://www.npmjs.com/package/@prisma/query-engine-wasm) version ~~`0.0.6`~~ `0.0.18`.

Closes #4121

### CODE REVIEW

File: query-engine/query-engine-wasm/build.sh
Comment (jkomyno) [inline query-engine/query-engine-wasm/build.sh:16]: Why `--dev` here?

Comment (SevInf) [reply query-engine/query-engine-wasm/build.sh:16]: Sorry. should've asked before commiting that. We should definitely go back to `--release` once we get it working, but until then it's conveneint to have readable stacktraces for panics. We can also revert it, i don't have strong feelings about it either way.

Comment (jkomyno) [reply query-engine/query-engine-wasm/build.sh:16]: No need to revert it right now :)
Was just looking for the rationale. Let's keep it this way until we get this to a working situation.

File: libs/crosstarget-utils/src/wasm/time.rs
Comment (jkomyno) [inline libs/crosstarget-utils/src/wasm/time.rs:45]: I'd rather silently dismiss the future error than `unwrap`-ing here.
What's the worst that could happen if `setTimeout` fails? Also, how likely is that it even happens?

File: query-engine/query-engine-wasm/build.sh
Comment (jkomyno) [inline query-engine/query-engine-wasm/build.sh:33]: Alexey and I have a few ideas on how to improve this `build.sh` script (also used by the automatic Github workflow that publishes new versions of `@prisma/query-engine-wasm`). We will however defer that work until after this is merged.

File: query-engine/connector-test-kit-rs/query-tests-setup/src/connector_tag/js/external_process.rs
Comment (jkomyno) [inline query-engine/connector-test-kit-rs/query-tests-setup/src/connector_tag/js/external_process.rs:102]: Nit: platform naming.

```suggestion
/// A Node.js process can die for a number of reasons, being one that any `panic!` occurring in Rust
```

File: query-engine/query-engine-wasm/src/wasm/tracer.rs
Comment (jkomyno) [inline query-engine/query-engine-wasm/src/wasm/tracer.rs:89]: Besides these 3 lines, the rest of the file is copied verbatim from `query-engine-node-api/src/tracer.rs`.

File: .github/workflows/formatting.yml
Comment (jkomyno) [inline .github/workflows/formatting.yml:29]:
```suggestion
        with:
          components: clippy
          targets: wasm32-unknown-unknown
```

File: libs/crosstarget-utils/src/native/spawn.rs
Comment (jkomyno) [inline libs/crosstarget-utils/src/native/spawn.rs:5]:
```suggestion
pub async fn spawn_if_possible<F>(future: F) -> Result<F::Output, SpawnError>
```

File: psl/psl-core/Cargo.toml
Comment (jkomyno) [inline psl/psl-core/Cargo.toml:27]: We most likely can get rid of this.

File: quaint/src/connector/metrics.rs
Comment (jkomyno) [inline quaint/src/connector/metrics.rs:49]: Should we use `.elapsed()` as a method name, like Instant does?

File: query-engine/connectors/sql-query-connector/src/database/js.rs
Comment (jkomyno) [inline query-engine/connectors/sql-query-connector/src/database/js.rs:17]: TODO: mention that this, for the time being, would cause symbol linking errors due to Napi.rs, as Miguel mentioned down below.

File: query-engine/core/src/executor/task.rs
Comment (jkomyno) [inline query-engine/core/src/executor/task.rs:13]: TODO: move to `crosstarget-utils`

File: query-engine/core/src/interactive_transactions/actor_manager.rs
Comment (jkomyno) [inline query-engine/core/src/interactive_transactions/actor_manager.rs:40]: Check who added this, if it wasn't us for this PR, then recover it.

File: query-engine/core/src/interactive_transactions/actors.rs
Comment (jkomyno) [inline query-engine/core/src/interactive_transactions/actors.rs:408]: Add an error trace in here.

File: query-engine/core/src/telemetry/capturing/mod.rs
Comment (jkomyno) [inline query-engine/core/src/telemetry/capturing/mod.rs:167]: Background info: this is used by the binary engine only, so we should place it in a different crate, and avoid importing it in the napi/wasm libquery case.

File: query-engine/driver-adapters/Cargo.toml
Comment (jkomyno) [inline query-engine/driver-adapters/Cargo.toml:16]:
```suggestion
```

File: query-engine/driver-adapters/Cargo.toml
Comment (jkomyno) [inline query-engine/driver-adapters/Cargo.toml:32]: TODO: import `quaint` only once, and leverage the resolver "2.0". cc @Weakky

File: query-engine/driver-adapters/src/lib.rs
Comment (jkomyno) [inline query-engine/driver-adapters/src/lib.rs:47]: TODO: avoid the top-level module imports, and use the fully-specialized types in the function type signatures, for explicitness

Comment (jkomyno) [reply query-engine/driver-adapters/src/lib.rs:47]: Note: this DOES NOT block the PR

File: query-engine/driver-adapters/src/lib.rs
Comment (jkomyno) [inline query-engine/driver-adapters/src/lib.rs:72]: Note: there's a `JsResult<T>` with the same name in `napi::result` / `wasm::result`; let's rename that one to `AdapterResult<T>` to avoid confusion.

File: query-engine/driver-adapters/src/send_future.rs
Comment (jkomyno) [inline query-engine/driver-adapters/src/send_future.rs:25]: TODO: rephrase comment to improve clarity

File: query-engine/driver-adapters/src/send_future.rs
Comment (jkomyno) [inline query-engine/driver-adapters/src/send_future.rs:10]: TODO:
- make it `///` for docs purposes
- remove "sent safely"
- highlight this is about `wasm-bindgen` limitations
- remind that we should probably don't require `Send` future for Wasm in the first place (this requires `#[async_trait(?)]` in e.g. the `Queryable` trait definition)

File: query-engine/driver-adapters/src/send_future.rs
Comment (jkomyno) [inline query-engine/driver-adapters/src/send_future.rs:12]:
```suggestion
pub struct UnsafeFuture<F: Future>(#[pin] pub F);
```

File: query-engine/driver-adapters/src/send_future.rs
Comment (jkomyno) [reply query-engine/driver-adapters/src/send_future.rs:25]: cc @miguelff

File: query-engine/driver-adapters/src/wasm/async_js_function.rs
Comment (jkomyno) [inline query-engine/driver-adapters/src/wasm/async_js_function.rs:25]:
```suggestion
    fn_: JsFunction,
```

File: query-engine/driver-adapters/src/wasm/js_object_extern.rs
Comment (jkomyno) [inline query-engine/driver-adapters/src/wasm/js_object_extern.rs:10]: TODO: add a comment explaining that this is a substitute for `js_sys::Reflect`

File: query-engine/driver-adapters/src/wasm/result.rs
Comment (jkomyno) [inline query-engine/driver-adapters/src/wasm/result.rs:24]: TODO: check whether this is duplicated, if so, don't duplicate it

File: query-engine/driver-adapters/src/wasm/result.rs
Comment (jkomyno) [inline query-engine/driver-adapters/src/wasm/result.rs:29]: TODO: add a comment that this is almost a duplicated of `napi::result::JsResult`, except that generic type bounds can't currently be applied conditionally based on the compilation target.

File: query-engine/query-engine-wasm/src/wasm/error.rs
Comment (jkomyno) [inline query-engine/query-engine-wasm/src/wasm/error.rs:3]: TODO: place this file in a new `query-engine-common` crate, to reduce duplicates.

File: query-engine/query-engine-wasm/src/wasm/tracer.rs
Comment (jkomyno) [reply query-engine/query-engine-wasm/src/wasm/tracer.rs:89]: TODO: use this implementation for Napi.rs as well, and move this to `query-engine-common`

File: query-engine/query-engine-wasm/src/wasm/error.rs
Comment (jkomyno) [reply query-engine/query-engine-wasm/src/wasm/error.rs:3]: See: https://github.com/prisma/prisma-engines/pull/4521.

**Review by SevInf — APPROVED** [review]
image

**Pull request merged**
